FlowCrypt/Functionality/Services/GoogleUserService.swift

```diff
@@ -229,6 +243,55 @@ extension GoogleUserService {
     }
 }
 
+// MARK: - Tokens
+extension GoogleUserService {
+    func getIdToken() async throws -> String {
+        guard let idToken = idToken else { throw(TokenError.missingToken) }
+
+        let decodedToken = try decode(idToken: idToken)
+
+        guard !decodedToken.isExpired else {
+            let (_, updatedToken) = try await performTokenRefresh()
+            return updatedToken
+        }
+
+        return idToken
+    }
+
+    private func decode(idToken: String) throws -> IdToken {
+        let components = idToken.components(separatedBy: ".")
+
+        guard components.count == 3 else { throw(TokenError.invalidJWTFormat) }
+
+        var decodedString = components[1]
+            .replacingOccurrences(of: "-", with: "+")
+            .replacingOccurrences(of: "_", with: "/")
+
+        while decodedString.utf16.count % 4 != 0 {
+            decodedString += "="
+        }
+
+        guard let decodedData = Data(base64Encoded: decodedString)
+        else { throw(TokenError.invalidBase64EncodedData) }
+
+        return try JSONDecoder().decode(IdToken.self, from: decodedData)
+    }
+
+    private func performTokenRefresh() async throws -> (accessToken: String, idToken: String) {
+        return try await withCheckedThrowingContinuation { continuation in
+            authorization?.authState.setNeedsTokenRefresh()
+            authorization?.authState.performAction { accessToken, idToken, error in
+                guard let accessToken = accessToken, let idToken = idToken else {
+                    let tokenError = error ?? AppErr.unexpected("Shouldn't happen because received nil error and nil token")
+                    return continuation.resume(throwing: tokenError)
+                }
+                let result = (accessToken, idToken)
+                return continuation.resume(with: .success(result))
+            }
+        }
+    }
```

Comment:
Does this work without user interaction?

Reply:
Yes, I tested it - token is automatically refreshed in background, without user interaction.